Preview server error since 0.7.2 in yarn workspace

Have tried to update to `mailing@^0.7.2` and unfortunately I'm getting this error when attempting to run `mailing preview`.

```
Server Error
Error: Cannot find module 'http'

This error happened while generating the page. Any console logs will be displayed in the terminal window.
Call Stack
webpackEmptyContext
file:///Users/willviles/Code/@Redbrain/packages/email/.mailing/.next/server/pages/index.js (22:10)
../../node_modules/axios/lib/adapters/http.js
webpack-internal:///./src/moduleManifest.js (1924:20)
__require2
webpack-internal:///./src/moduleManifest.js (34:56)
getDefaultAdapter
webpack-internal:///./src/moduleManifest.js (2240:27)
../../node_modules/axios/lib/defaults/index.js
webpack-internal:///./src/moduleManifest.js (2259:22)
__require2
webpack-internal:///./src/moduleManifest.js (34:56)
../../node_modules/axios/lib/core/transformData.js
webpack-internal:///./src/moduleManifest.js (2338:24)
__require2
webpack-internal:///./src/moduleManifest.js (34:56)
../../node_modules/axios/lib/core/dispatchRequest.js
webpack-internal:///./src/moduleManifest.js (2362:29)
__require2
webpack-internal:///./src/moduleManifest.js (34:56)
```

There are no errors in a vanilla project, however I'm using Mailing within a yarn workspace, which already contains a next.js app. 

Pinning to `mailing@0.7.1` and adding a the following nohoist rules to my root `package.json` ensures the error disappears:

```json
{
  "workspaces": {
    "nohoist": [
      "**/mailing",
      "**/mailing-core"
    ]
  }
}
```

--- a/.changeset/pre.json
+++ b/.changeset/pre.json
@@ -1,0 +1,12 @@
+{
+  "mode": "pre",
+  "tag": "next",
+  "initialVersions": {
+    "mailing": "0.7.4",
+    "mailing-core": "0.7.4",
+    "web": "0.3.4"
+  },
+  "changesets": [
+    "shiny-dolphins-trade"
+  ]
+}

--- a/.changeset/shiny-dolphins-trade.md
+++ b/.changeset/shiny-dolphins-trade.md
@@ -1,0 +1,5 @@
+---
+"mailing": patch
+---
+
+Fix #163 by updating the "externals" option passed to esbuild to include more possible node_modules directories.

--- a/packages/cli/src/commands/preview/server/__test__/livereload.test.ts
+++ b/packages/cli/src/commands/preview/server/__test__/livereload.test.ts
@@ -54,7 +54,7 @@ describe("livereload", () => {
         fetchClock(startClock),
         fetchClock(startClock),
       ]);
-      expect(Date.now() - start).toBeLessThan(1000);
+      expect(Date.now() - start).toBeLessThan(3000);
       expect(Date.now() - start).toBeGreaterThan(100);
       done();
     }, 0);

--- a/packages/cli/src/commands/preview/server/__test__/livereload.test.ts
+++ b/packages/cli/src/commands/preview/server/__test__/livereload.test.ts
@@ -43,7 +43,7 @@ describe("livereload", () => {
 
   it("returns pollers immediatly when vectorClock=0", async () => {
     await Promise.all([fetchClock(0), fetchClock(0), fetchClock(0)]);
-    expect(Date.now() - start).toBeLessThan(200);
+    expect(Date.now() - start).toBeLessThan(2000);
   });
 
   it("returns pollers immediatly when vectorClock=0", (done) => {

--- a/packages/cli/src/commands/preview/server/__test__/setup.test.ts
+++ b/packages/cli/src/commands/preview/server/__test__/setup.test.ts
@@ -3,7 +3,6 @@ import path from "path";
 import {
   packageJsonVersionsMatch,
   bootstrapMailingDir,
-  getNodeModulesDirs,
   COMPONENT_FILE_REGEXP,
 } from "../setup";
 import childProcess from "child_process";
@@ -117,18 +116,6 @@ describe("setup", () => {
 
       await bootstrapMailingDir();
       expect(fsExtra.copy).not.toHaveBeenCalled();
-    });
-  });
-
-  describe("getNodeModulesDirs", () => {
-    it("should return an array with all the node_modules folders", () => {
-      jest.spyOn(path, "resolve").mockImplementation(() => "/a/b/c");
-      expect(getNodeModulesDirs()).toEqual([
-        "../../../node_modules",
-        "../../node_modules",
-        "../node_modules",
-        "./node_modules",
-      ]);
     });
   });
 

--- a/packages/cli/src/commands/preview/server/__test__/setup.test.ts
+++ b/packages/cli/src/commands/preview/server/__test__/setup.test.ts
@@ -1,5 +1,4 @@
 import fsExtra from "fs-extra";
-import path from "path";
 import {
   packageJsonVersionsMatch,
   bootstrapMailingDir,

--- a/packages/cli/src/commands/preview/server/__test__/setup.test.ts
+++ b/packages/cli/src/commands/preview/server/__test__/setup.test.ts
@@ -1,7 +1,9 @@
 import fsExtra from "fs-extra";
+import path from "path";
 import {
   packageJsonVersionsMatch,
   bootstrapMailingDir,
+  getNodeModulesDirs,
   COMPONENT_FILE_REGEXP,
 } from "../setup";
 import childProcess from "child_process";
@@ -25,101 +27,119 @@ function mockReadJSONSyncToThrowErrno() {
   });
 }
 
-describe("packageJsonVersionsMatch", () => {
-  it("should return false if .mailing package.json does not exist", async () => {
-    mockReadJSONSyncToThrowErrno();
-    const match = await packageJsonVersionsMatch();
-    expect(match).toBe(false);
+describe("setup", () => {
+  beforeEach(() => {
+    jest.restoreAllMocks();
   });
 
-  it("should return true if .mailing package.json matches the cli version", async () => {
-    mockPackageJsonVersionToMatch();
+  describe("packageJsonVersionsMatch", () => {
+    it("should return false if .mailing package.json does not exist", async () => {
+      mockReadJSONSyncToThrowErrno();
+      const match = await packageJsonVersionsMatch();
+      expect(match).toBe(false);
+    });
 
-    const match = await packageJsonVersionsMatch();
-    expect(match).toBe(true);
+    it("should return true if .mailing package.json matches the cli version", async () => {
+      mockPackageJsonVersionToMatch();
+
+      const match = await packageJsonVersionsMatch();
+      expect(match).toBe(true);
+    });
+
+    it("should return false if .mailing package.json does not match the cli version", async () => {
+      // mock the .mailing/package.json version
+      jest
+        .spyOn(fsExtra, "readJSONSync")
+        .mockImplementation(() => ({ version: "1.2.3" }));
+
+      // mock the npx mailing version
+      jest.spyOn(childProcess, "execSync").mockImplementation(() => "4.5.6");
+
+      const match = await packageJsonVersionsMatch();
+      expect(match).toBe(false);
+    });
   });
 
-  it("should return false if .mailing package.json does not match the cli version", async () => {
-    // mock the .mailing/package.json version
-    jest
-      .spyOn(fsExtra, "readJSONSync")
-      .mockImplementation(() => ({ version: "1.2.3" }));
+  describe("bootstrapMailingDir", () => {
+    it("should rm, mkdir, and copy", async () => {
+      mockReadJSONSyncToThrowErrno();
 
-    // mock the npx mailing version
-    jest.spyOn(childProcess, "execSync").mockImplementation(() => "4.5.6");
+      const match = await packageJsonVersionsMatch();
+      expect(match).toBe(false);
 
-    const match = await packageJsonVersionsMatch();
-    expect(match).toBe(false);
-  });
-});
+      // setup mocks for copy, rm, and mkdir
+      const copyMock = jest.fn(() => true);
+      jest.spyOn(fsExtra, "copy").mockImplementation(copyMock);
 
-describe("bootstrapMailingDir", () => {
-  it("should rm, mkdir, and copy", async () => {
-    mockReadJSONSyncToThrowErrno();
+      const rmMock = jest.fn();
+      jest.spyOn(fsExtra, "rm").mockImplementation(rmMock);
 
-    const match = await packageJsonVersionsMatch();
-    expect(match).toBe(false);
+      const mkdirMock = jest.fn();
+      jest.spyOn(fsExtra, "mkdir").mockImplementation(mkdirMock);
 
-    // setup mocks for copy, rm, and mkdir
-    const copyMock = jest.fn(() => true);
-    jest.spyOn(fsExtra, "copy").mockImplementation(copyMock);
+      // don't actually write a .gitignore while the test is running
+      const writeFileMock = jest.fn();
+      jest.spyOn(fsExtra, "writeFile").mockImplementation(writeFileMock);
 
-    const rmMock = jest.fn();
-    jest.spyOn(fsExtra, "rm").mockImplementation(rmMock);
+      await bootstrapMailingDir();
 
-    const mkdirMock = jest.fn();
-    jest.spyOn(fsExtra, "mkdir").mockImplementation(mkdirMock);
+      // copy preview directory to ./.mailing
+      expect(copyMock).toHaveBeenCalledTimes(1);
+      const copyCall: any[] = copyMock.mock.calls[0];
+      expect(copyCall[0]).toMatch(
+        new RegExp("/mailing/packages/cli/src/commands/preview")
+      );
+      expect(copyCall[1]).toBe("./.mailing");
+      expect(copyCall[2].recursive).toBe(true);
 
-    // don't actually write a .gitignore while the test is running
-    const writeFileMock = jest.fn();
-    jest.spyOn(fsExtra, "writeFile").mockImplementation(writeFileMock);
+      // rm -f ./.mailing
+      expect(rmMock).toHaveBeenCalledTimes(2);
+      const rmDotMailingCall: any[] = rmMock.mock.calls[0];
+      expect(rmDotMailingCall[0]).toBe("./.mailing");
+      expect(rmDotMailingCall[1]).toEqual({ force: true, recursive: true });
 
-    await bootstrapMailingDir();
+      // rm -f ./.mailing/src/emails
+      const rmDotMailingSrcEmailsCall: any[] = rmMock.mock.calls[1];
+      expect(rmDotMailingSrcEmailsCall[0]).toBe("./.mailing/src/emails");
+      expect(rmDotMailingCall[1]).toEqual({ force: true, recursive: true });
 
-    // copy preview directory to ./.mailing
-    expect(copyMock).toHaveBeenCalledTimes(1);
-    const copyCall: any[] = copyMock.mock.calls[0];
-    expect(copyCall[0]).toMatch(
-      new RegExp("/mailing/packages/cli/src/commands/preview")
-    );
-    expect(copyCall[1]).toBe("./.mailing");
-    expect(copyCall[2].recursive).toBe(true);
+      // mkdir ./.mailing
+      expect(mkdirMock).toHaveBeenCalledTimes(1);
+      const mkdirCall: any[] = mkdirMock.mock.calls[0];
+      expect(mkdirCall[0]).toBe("./.mailing");
+      expect(mkdirCall[1]).toEqual({ recursive: true });
+    });
 
-    // rm -f ./.mailing
-    expect(rmMock).toHaveBeenCalledTimes(2);
-    const rmDotMailingCall: any[] = rmMock.mock.calls[0];
-    expect(rmDotMailingCall[0]).toBe("./.mailing");
-    expect(rmDotMailingCall[1]).toEqual({ force: true, recursive: true });
+    it("should return early if packageJsonVersionsMatch is true", async () => {
+      mockPackageJsonVersionToMatch();
 
-    // rm -f ./.mailing/src/emails
-    const rmDotMailingSrcEmailsCall: any[] = rmMock.mock.calls[1];
-    expect(rmDotMailingSrcEmailsCall[0]).toBe("./.mailing/src/emails");
-    expect(rmDotMailingCall[1]).toEqual({ force: true, recursive: true });
+      jest.spyOn(fsExtra, "copy").mockImplementation(jest.fn);
 
-    // mkdir ./.mailing
-    expect(mkdirMock).toHaveBeenCalledTimes(1);
-    const mkdirCall: any[] = mkdirMock.mock.calls[0];
-    expect(mkdirCall[0]).toBe("./.mailing");
-    expect(mkdirCall[1]).toEqual({ recursive: true });
+      await bootstrapMailingDir();
+      expect(fsExtra.copy).not.toHaveBeenCalled();
+    });
   });
 
-  it("should return early if packageJsonVersionsMatch is true", async () => {
-    mockPackageJsonVersionToMatch();
-
-    jest.spyOn(fsExtra, "copy").mockImplementation(jest.fn);
-
-    await bootstrapMailingDir();
-    expect(fsExtra.copy).not.toHaveBeenCalled();
+  describe("getNodeModulesDirs", () => {
+    it("should return an array with all the node_modules folders", () => {
+      jest.spyOn(path, "resolve").mockImplementation(() => "/a/b/c");
+      expect(getNodeModulesDirs()).toEqual([
+        "../../../node_modules",
+        "../../node_modules",
+        "../node_modules",
+        "./node_modules",
+      ]);
+    });
   });
-});
 
-describe("SOURCE_FILE_REGEXP", () => {
-  it("should match jsx and tsx without matching js and ts files", () => {
-    expect(COMPONENT_FILE_REGEXP.test("file.ts")).toBe(false);
-    expect(COMPONENT_FILE_REGEXP.test("file.js")).toBe(false);
-    expect(COMPONENT_FILE_REGEXP.test("file.tsx")).toBe(true);
-    expect(COMPONENT_FILE_REGEXP.test("file.jsx")).toBe(true);
-    expect(COMPONENT_FILE_REGEXP.test("bad file name.jsx")).toBe(false);
-    expect(COMPONENT_FILE_REGEXP.test("bad file name.tsx")).toBe(false);
+  describe("SOURCE_FILE_REGEXP", () => {
+    it("should match jsx and tsx without matching js and ts files", () => {
+      expect(COMPONENT_FILE_REGEXP.test("file.ts")).toBe(false);
+      expect(COMPONENT_FILE_REGEXP.test("file.js")).toBe(false);
+      expect(COMPONENT_FILE_REGEXP.test("file.tsx")).toBe(true);
+      expect(COMPONENT_FILE_REGEXP.test("file.jsx")).toBe(true);
+      expect(COMPONENT_FILE_REGEXP.test("bad file name.jsx")).toBe(false);
+      expect(COMPONENT_FILE_REGEXP.test("bad file name.tsx")).toBe(false);
+    });
   });
 });

--- a/packages/cli/src/commands/preview/server/setup.ts
+++ b/packages/cli/src/commands/preview/server/setup.ts
@@ -192,6 +192,8 @@ async function buildManifest(
 ) {
   const buildOutdir = ".mailing/src";
 
+  const globalNodeModulesDirectory = execSync("npm -g root").toString();
+
   const buildOpts: BuildOptions = {
     entryPoints: [manifestPath],
     outdir: buildOutdir,
@@ -199,7 +201,9 @@ async function buildManifest(
     bundle: true,
     format: "esm",
     jsx: "preserve",
-    external: getNodeModulesDirsFrom("."),
+    external: getNodeModulesDirsFrom(".").concat(
+      getNodeModulesDirsFrom(globalNodeModulesDirectory)
+    ),
   };
 
   if ("node" === buildType) {

--- a/packages/cli/src/commands/preview/server/setup.ts
+++ b/packages/cli/src/commands/preview/server/setup.ts
@@ -1,4 +1,4 @@
-import { relative, resolve } from "path";
+import { relative, resolve, normalize, sep } from "path";
 import { execSync } from "child_process";
 import {
   copy,
@@ -185,6 +185,21 @@ export async function bootstrapMailingDir() {
   }
 }
 
+export function getNodeModulesDirs() {
+  const nodeModulesDirs = [];
+
+  const pathDepth = resolve(".").split(sep).length - 1;
+  let i = pathDepth;
+
+  do {
+    nodeModulesDirs.push(normalize(`..${sep}`.repeat(i) + "node_modules"));
+  } while (--i);
+
+  nodeModulesDirs.push("./node_modules");
+
+  return nodeModulesDirs;
+}
+
 async function buildManifest(
   buildType: "node" | "browser",
   manifestPath: string
@@ -198,13 +213,7 @@ async function buildManifest(
     bundle: true,
     format: "esm",
     jsx: "preserve",
-    external: [
-      "./node_modules/*",
-      "../node_modules/*",
-      "../../node_modules/*",
-      "../../../node_modules/*",
-      "../../../../node_modules/*",
-    ],
+    external: getNodeModulesDirs(),
   };
 
   if ("node" === buildType) {

--- a/packages/cli/src/commands/preview/server/setup.ts
+++ b/packages/cli/src/commands/preview/server/setup.ts
@@ -1,4 +1,4 @@
-import { relative, resolve, normalize, sep } from "path";
+import { relative, resolve } from "path";
 import { execSync } from "child_process";
 import {
   copy,
@@ -14,6 +14,7 @@ import {
 
 import { debug, log } from "../../../util/log";
 import { build, BuildOptions } from "esbuild";
+import { getNodeModulesDirsFrom } from "../../util/getNodeModulesDirsFrom";
 
 export const COMPONENT_FILE_REGEXP = /^[^\s-]+\.[tj]sx$/; // no spaces, .jsx or .tsx
 
@@ -185,21 +186,6 @@ export async function bootstrapMailingDir() {
   }
 }
 
-export function getNodeModulesDirs() {
-  const nodeModulesDirs = [];
-
-  const pathDepth = resolve(".").split(sep).length - 1;
-  let i = pathDepth;
-
-  do {
-    nodeModulesDirs.push(normalize(`..${sep}`.repeat(i) + "node_modules"));
-  } while (--i);
-
-  nodeModulesDirs.push("./node_modules");
-
-  return nodeModulesDirs;
-}
-
 async function buildManifest(
   buildType: "node" | "browser",
   manifestPath: string
@@ -213,7 +199,7 @@ async function buildManifest(
     bundle: true,
     format: "esm",
     jsx: "preserve",
-    external: getNodeModulesDirs(),
+    external: getNodeModulesDirsFrom("."),
   };
 
   if ("node" === buildType) {

--- a/packages/cli/src/commands/preview/server/setup.ts
+++ b/packages/cli/src/commands/preview/server/setup.ts
@@ -1,6 +1,5 @@
 import { relative, resolve } from "path";
 import { execSync } from "child_process";
-import pkg from "../../../../package.json";
 import {
   copy,
   mkdir,
@@ -200,8 +199,11 @@ async function buildManifest(
     format: "esm",
     jsx: "preserve",
     external: [
-      ...Object.keys(pkg.dependencies || {}),
-      ...Object.keys(pkg.peerDependencies || {}),
+      "./node_modules/*",
+      "../node_modules/*",
+      "../../node_modules/*",
+      "../../../node_modules/*",
+      "../../../../node_modules/*",
     ],
   };
 
@@ -212,6 +214,8 @@ async function buildManifest(
     buildOpts.platform = "browser";
     buildOpts.target = "esnext";
   }
+
+  debug("invoking esbuild with buildOpts: ", JSON.stringify(buildOpts));
 
   // build the manifest
   debug(`bundling ${buildType} manifest for ${manifestPath}...`);

--- a/packages/cli/src/commands/util/__test__/getNodeModulesDirsFrom.test.ts
+++ b/packages/cli/src/commands/util/__test__/getNodeModulesDirsFrom.test.ts
@@ -1,0 +1,13 @@
+import { sep } from "path";
+import { getNodeModulesDirsFrom } from "../getNodeModulesDirsFrom";
+
+describe("getNodeModulesDirsFrom", () => {
+  it("should return an array with all the node_modules folders", () => {
+    const output = getNodeModulesDirsFrom(".");
+    expect(output.length).toBeGreaterThan(0);
+
+    // for esbuild to look in the right place, it must end in "/*"
+    const regexp = new RegExp(`${sep}node_modules${sep}\\*$`);
+    expect(output[0]).toMatch(regexp);
+  });
+});

--- a/packages/cli/src/commands/util/getNodeModulesDirsFrom.ts
+++ b/packages/cli/src/commands/util/getNodeModulesDirsFrom.ts
@@ -1,0 +1,17 @@
+import { resolve, join, sep } from "path";
+
+// traversing up from startPath to the root, for each directory append a "node_modules" directory
+export function getNodeModulesDirsFrom(startPath: string) {
+  const nodeModulesDirs = [];
+
+  const pathDepth = resolve(startPath).split(sep).length;
+  let i = pathDepth;
+
+  do {
+    nodeModulesDirs.push(
+      resolve(join(startPath, `..${sep}`.repeat(i - 1), "node_modules", "*"))
+    );
+  } while (--i);
+
+  return nodeModulesDirs;
+}

